### PR TITLE
Fix calldata encoding for using gelato keepers

### PR DESCRIPTION
**Pull request by shuklaayush**
- Also changed all imports to absolute to make it compatible with hardhat

### CODE REVIEW

**Review by petrovska-petro — DISMISSED** [review]
Compatibility for both keepers has being updated and now we can leverage either cl keeper or gelato as we pleased by updating the `keeper` address.

Also, in the processing of rewards has being added extra checkups to ensure that when amount is zero does not fall into that code block for executing any method with zero value to adapt to the latest tcl strategy which locks 100% of aura and sells 100% of bal for usdc

File: src/aura/AuraAvatarTwoToken.sol [hunk 1]
```diff
@@ -587,47 +595,79 @@ contract AuraAvatarTwoToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtils,
...
-        USDC.safeTransfer(ownerCached, totalUsdcEarned);
+        if (totalUsdcEarned > 0) {
+            USDC.safeTransfer(ownerCached, totalUsdcEarned);
+        }
 
         // 3. Deposit remaining BAL to 80BAL-20ETH BPT
-        depositBalToBpt(totalBal - balForUsdc);
+        uint256 balToBpt = totalBal - balForUsdc;
+        if (balToBpt > 0) {
+            depositBalToBpt(totalBal - balForUsdc);
```
Comment (sajanrajdev) [inline src/aura/AuraAvatarTwoToken.sol:642]: Why not using `balToBpt ` as the parameter here?

File: src/aura/AuraAvatarTwoToken.sol [hunk 2]
```diff
@@ -587,47 +595,79 @@ contract AuraAvatarTwoToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtils,
...
+        }
 
         // 4. Swap BPT for auraBAL or lock
-        swapBptForAuraBal(BPT_80BAL_20WETH.balanceOf(address(this)));
+        uint256 bptToAuraBal = BPT_80BAL_20WETH.balanceOf(address(this));
+        if (bptToAuraBal > 0) {
+            swapBptForAuraBal(bptToAuraBal);
```
Comment (sajanrajdev) [inline src/aura/AuraAvatarTwoToken.sol:648]: Could there be a case in which the previous block (depositValToBpt) runs and the bpt resulting balance is 0? If not, wouldn't it make sense to move the following to the `if (balToBpt > 0) {}` block?
```
uint256 bptToAuraBal = BPT_80BAL_20WETH.balanceOf(address(this));
swapBptForAuraBal(bptToAuraBal);
```

File: src/aura/AuraAvatarTwoToken.sol [hunk 3]
```diff
@@ -587,47 +595,79 @@ contract AuraAvatarTwoToken is BaseAvatar, PausableUpgradeable, AuraAvatarUtils,
...
-        uint256 bauraBalBefore = BAURABAL.balanceOf(ownerCached);
-        BAURABAL.depositFor(ownerCached, AURABAL.balanceOf(address(this)));
-        uint256 bauraBalAfter = BAURABAL.balanceOf(ownerCached);
+        uint256 auraBalBalance = AURABAL.balanceOf(address(this));
+        uint256 bAuraBalEarned;
+        if (auraBalBalance > 0) {
+            uint256 bAuraBalBefore = BAURABAL.balanceOf(ownerCached);
+            BAURABAL.depositFor(ownerCached, auraBalBalance);
+            bAuraBalEarned = BAURABAL.balanceOf(ownerCached) - bAuraBalBefore;
+        }
```
Comment (sajanrajdev) [inline src/aura/AuraAvatarTwoToken.sol:658]: Same with this block, wouldn't it make sense to merge steps 3, 4 and 5 into the same block and use only one balance check? In the end, `balToBpt` will be more than 0 if, only if there was any BAL to begin with and we didn't sell it all for USDC.

**Review by sajanrajdev — COMMENTED** [review]
- Keeper functions were generalized to work with both Gelato and CL keepers according to their specific requirements (Still, only when not paused)
- Operations in the main processing function will only execute if the BPS for those operations are more than 0
- Tests were adapted to cover the above

File: src/aura/AuraAvatarTwoToken.sol (hunk 1 above)
Comment (shuklaayush) [reply src/aura/AuraAvatarTwoToken.sol:642]: Oops, fixed

File: src/aura/AuraAvatarTwoToken.sol (hunk 2 above)
Comment (shuklaayush) [reply src/aura/AuraAvatarTwoToken.sol:648]: Good point. Yep, that's cleaner

**Review by sajanrajdev — DISMISSED** [review]
LGTM

image

**Review by sajanrajdev — APPROVED** [review]

**Pull request merged**